components/dashboards-web-component/src/utils/WidgetPubSubHub.js

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+export default class WidgetPubSubHub {
+    constructor(glEventHub) {
+        this.glEventHub = glEventHub;
+        this.lastPublishedMessages = {};
+    }
+
+    /**
+     * Publishes a message to subscribers.
+     * @param {string} publisherId ID of the message publisher
+     * @param {object} message publishing message
+     */
+    publish(publisherId, message) {
+        this.lastPublishedMessages[publisherId] = message;
+        this.glEventHub.emit(publisherId, message);
```

Comment:
Shouldn't these lines be interchanged ? What if the publishing process breaks in the middle? We will have to rollback the map if we used this way.

Reply:
The only reason to break `emit` is an error happens in the `listenerCallback` function which is provided by the widget author.
If _message1_ causes _subscriber1_'s `listenerCallback` to throw an error, it doesn't mean that _subscriber2_ will behave the same (throw an error). So we shouldn't prevent _subscriber2_ from getting _message1_.

If we really want to protect this code from error throwing/happening in subscriber's `listenerCallback` functions, then this line should be wrapped in a `try-catch` block.

Reply:
Is emit coupled with ListernerCallback? It should not be. Because publisher should emit to the channel no matter on what happens at subscriber end.

Reply:
publisher and subscriber is decoupled by the GoldenLayout EventHub. `glEventHub.emit()` will ultimately invoke the `listenerCallback` provided previously by the subscriber.
See `emit` implementation https://github.com/golden-layout/golden-layout/blob/v1.5.9/src/js/utils/EventEmitter.js#L46

Reply:
glEventHub.emit() will only notify the callbacks and it will not invoke the callbacks.